lib/app/models/sequence_requirement.rb

```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module Inferno
+  module Models
+    class SequenceRequirement
+      include DataMapper::Resource
+      property :id, String, key: true, default: proc { SecureRandom.uuid }
+      property :name, String
```

Comment:
Added, though not sure if I'm supposed to pass in the testing_instance_id when creating or if it'll figure it out.

Reply:
In `add_sequence_requirements`, rather than a `new` and a `push`, use a create and pass in the testing instance.
```ruby
requirements.each do |requirement, texts|
  SequenceRequirement.create(
    name: requirement,
    value: '',
    label: texts[:label],
    description: texts[:description],
    testing_instance: self
    # "testing_instance_id: id" should work if this doesn't
  )
end
```

You also shouldn't need the final `save!` since the `TestingInstance` itself isn't being changed.

Reply:
I think this is working with add_sequence_requirements, but its causing errors when I try to add requirements in get_requirement_value and set_requirement_value. Trying it this way and running rake gives a ton of errors.

Reply:
I was also getting similar errors around the unique constraints on `name` and `testing_instance_id`

e.g. `DataObjects::IntegrityError: UNIQUE constraint failed: inferno_models_sequence_requirements.name, inferno_models_sequence_requirements.testing_instance_id`

I was able to resolve it by adding another check before creating and using `#first`.  Interestingly using either:

`sequence_requirements.find { |req| req.name == requirement_name.to_s }`
or
`sequence_requirements.find { |req| req.name == requirement_name.to_s && req.testing_instance_id == id }`

was not returning up to date results, while

`sequence_requirements.first(name: requirement_name.to_s, testing_instance_id: id)` was (you can test by just swapping back the existing).

This is leading me to believe that using `#find` might not be the most reliable way to search through the associated tables (maybe there is an "update references" or something method, but I think using `#first` seems clearer and its a DataMapper method built for this purpose).


Either way seems to be working now!

Reply:
These errors are why we needed the uniqueness constraint. We don't want multiple instances of these values to exist. Using `.first` and then creating or updating introduces a race condition (the record could be created in between when you check and when you create).

`DataMapper` has a [`first_or_create` method](https://www.rubydoc.info/github/datamapper/dm-core/DataMapper%2FModel:first_or_create) that should prevent this race condition from occurring. Here's [an example using it](https://stackoverflow.com/questions/6509788/datamapper-first-or-create-always-set-certain-field).

Reply:
Awesome, good suggestion!  Changed to use `#first_or_create`.  Added a helper method around it to contain the defaults and logic for conditions and defaults logic.